pennylane/clusters.py

```diff
@@ -0,0 +1,455 @@
+
+    or looped over:
+
+    >>> [i.num_wires for i in qnodes]
+    [2, 2, 2]
+
+    To evaluate a QNodeCluster, simply call the cluster, passing the parameters
+    as required by the constituent QNode. For example, consider the
+    following two QNodes with the same signature:
+
+    .. code-block:: python3
+
+        dev1 = qml.device("default.qubit", wires=1)
+        dev2 = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev1)
+        def qnode1(x, y):
+            qml.RX(x, wires=0)
+            qml.RY(y, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        @qml.qnode(dev2)
+        def qnode2(x, y):
+            qml.Hadamard(wires=0)
+            qml.RX(x, wires=0)
+            qml.RY(y, wires=1)
+            qml.CNOT(wires=[0, 1])
+            return qml.expval(qml.PauliZ(0)), qml.var(qml.PauliZ(1))
+
+    Creating a QNodeCluster,
+
+    >>> qnodes = qml.QNodeCluster([qnode1, qnode2])
+
+    We can evaluate this QNode cluster directly:
+
+    >>> qnodes(0.5643, -0.45)
+    [ 7.60844651e-01 -5.55111512e-17  1.00000000e+00]
+
+    where the results from each QNode have been flattened and concatenated
+    into a single one-dimensional list.
+    """
+
+    def __init__(self, qnodes=None):
+        self.qnodes = []
+        self.extend(qnodes or [])
+
+    @property
+    def interface(self):
+        """str, None: automatic differentiation interface used by the cluster, if any"""
+        if not self.qnodes:
+            return None
+
+        return self.qnodes[0].interface
+
+    def append(self, qnode):
+        """Appends a QNode to the cluster. The appended QNode *must* have the same
+        interface as the QNode cluster."""
+        self.extend([qnode])
+
+    def extend(self, qnodes):
+        """Extends the cluster by a list of QNodes. The appended QNodes *must* have the same
+        interface as the QNode cluster."""
+        if not all(i.interface == qnodes[0].interface for i in qnodes):
+            raise ValueError("Provided QNodes do not all use the same interface")
+
+        if self.qnodes and (qnodes[0].interface != self.interface):
+            raise ValueError(
+                "Interface mismatch. Provided QNodes use the {} interface, "
+                "QNode cluster uses the {} interface".format(qnodes[0].interface, self.interface)
+            )
+
+        self.qnodes.extend(qnodes)
+
+    def __call__(self, *args, **kwargs):
+        results = []
+
+        for q in self.qnodes:
+            # TODO: allow asynchronous QNode evaluation here
+            results.append(q(*args, **kwargs))
+
+        if self.interface == "tf":
+            import tensorflow as tf
+
+            return tf.stack(results)
+
+        if self.interface == "torch":
+            import torch
+
+            return torch.stack(results, dim=0)
+
+        if self.interface in ("autograd", "numpy"):
+            from autograd import numpy as np
+
+            return np.vstack(results)
```

Comment:
This part of ``__call__`` seems to be logically different from the first ``for`` loop (more like ``process_results()`` whereas the previous is more like ``process_qnodes()``). It could be organized into a separate method as well such that the logic is better separated.

(When reading for the first time, I kept looking for differences along the if statements.)

Reply:
done!